wordpress-mobile/WordPress-Android
# Blaze Promote: web view refactoring

This PR implements refactor a good portion of the Blaze Promote webview/fragment to use UI states. This PR is very similar to #18844 

Changes include
-The view moves from `Preparing` to `Loading` to `Loaded` states. The loading indicator is shown while in the preparing and loading states. It gets removed on call from the webview that the page has loaded.

-Show a loading indicator, the view will now show an error view if account details are not present, there is no network connectivity, or the webview throws an error. I included a "retry" button, not sure if it should be there for non network connectivity errors. wdyt? FYI @osullivanchris 

Loading Before | Loading After
-- | --
<video src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/d67752f6-de71-4ed6-b045-32cc6b25d711"> | <video src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/a6521682-fd88-4b34-a901-7f0c71e0762b">

No Network Dark | No Network Light
-- | --
<img height="400" width="200" src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/a7154884-3d92-4883-8fcc-6eb78d0eb89c"> | <img height="400" width="200" src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/8814bc62-d547-483e-b331-5f59db59760a">

Error Dark | Error Light
-- | --
<img height="400" width="200" src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/12bc69dc-1473-4847-a529-f2a5e89d3ae9"> | <img height="400" width="200" src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/1a65345e-151c-4383-8b9c-6277f3ae2390">

### Merge Instructions
- Ensure #18844 has been merged
- Remove Not Ready for merge label
- Merge as normal

### To test:
**Setup**
- Install and launch the app
- Navigate to Me -> App Settings -> Privacy Setting
- Enable collection information
- Navigate to Me -> App Settings -> Debug Setting
- Enable Blaze
- Restart the app

**Viewing Promote Blaze Flow**
- Navigate to a site that has blaze campaigns (use the test creds or create a blaze campaign first)
- Navigate to My Site tab
- Tap on the Create Campaign link within the Blaze Card to launch the Promote with Blaze Flow
- ✅ Verify the Loading indicator is shown while the page is loading
- ✅ Verify the Loading indicator goes away when the page is loaded

**Error view shows when there is no internet**
- Navigate to a site that has blaze campaigns (use the test creds or create a blaze campaign first)
- Navigate to My Site tab > Menu tab
- Tap on the Blaze menu item
- Verify the campaign list view has launched
- **STOP** - Turn on airplane mode
- Tap on the FAB button 
- ✅ Verify the Blaze view is launched, the header is visible, and the No Internet Error view is shown

**Error view shows when there is no account creds/or the webview is throws an error**
- Download the branch
- In fun `validateAndPostErrorIfNeeded` comment out the if statement and the return false
- Build, install, and run
- Navigate to My SIte tab 
- Tap on the Create Campaign link within the Blaze Card to launch the Promote with Blaze Flow
✅ Verify the generic error view is shown  

**Test Dashboard Entry Point**
- Navigate to My SIte tab
- Tap on the Promote With Blaze card
- If the blaze overlay is shown, then tap on the Blaze a Post Now button within the overlay. The blaze overlay will only show 1X
- ✅ Verify you are launched into the Blaze flow post lists
- ✅ Verify logs contain:  🔵 Tracked: blaze_flow_started, Properties: {"source":"dashboard_card"}
- Tap the cancel button in the header
- ✅ Verify you are taken out of the flow 
- ✅ Verify logs contain: 🔵 Tracked: blaze_flow_canceled, Properties: {"source":"dashboard_card","current_step":"posts_list"}
- Enter the blaze flow again
- Select a post from the list and tap Promote
- Tap the cancel button in the header
- ✅ Verify you are taken out of the flow 
---------------------------
- Enter the flow again
- Select a post from the list and tap Promote
- Fill out the fields (Change the title and snippet to DO NOT APPROVE - TESTING)
- Fill out the Payment information and submit
- Note that the cancel button is un-clickable
- Wait for the process to complete
-  ✅ Verify the header action button changed to "done"
- Tap the Header "Done" button

**Test Post Entry**
- Navigate to My SIte tab
- Tap on the Posts menu item or ribbon action
- Tap on More > Promote with Blaze
- ✅ Verify you are launched into the Blaze flow for the post you selected
- Tap the cancel button in the header
- ✅ Verify you are taken out of the flow 


## Regression Notes
1. Potential unintended areas of impact
The blaze flow does not work as expected

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
N/A

**PR submission checklist:**

**UI Changes testing checklist:**
